Getting Started with Rails > 5.2 The first form (#232)
campuscode/rails-guides-pt-BR: pt-BR/getting_started.md

Ao atualizar <http://localhost:3000/articles/new> você verá que a página tem um título. A rota, o *controller*, a *action* e a *view* estão funcionando harmoniosamente! É hora de criar o formulário para um novo artigo.

### O primeiro formulário

> Para criar um formulário com este template, você vai usar um construtor de formulário (*form builder*). O *form builder* principal do Rails é disponibilizado por um *helper* chamado `form_with`. Para utilizá-lo, adicione o código abaixo no arquivo `app/views/articles/new.html.erb`:

Reviewer: ```suggestion
Para criar um formulário com este template, você vai usar um construtor de formulário (*form builder*). O construtor de formulários do Rails é disponibilizado por um método *helper* chamado `form_with`. Para utilizá-lo, adicione o código abaixo no arquivo `app/views/articles/new.html.erb`:
```

O primeiro form builder é o termo técnico o segundo acho que podemos explicar com mais detalhes o significado dele, o que acha?

> ```html+erb

Reviewer: @campuscode/guia-rails-team acho que seria legal modificar esse código para mostrar o formulário em 'pt-BR' vocês acham que isso seria prejudicial?

O que você acha @cleicar?

Reviewer 2: @HenriqueMorato você diz para mudarmos para algo como `<%= form.text_field :title, 'Título' %>` ?

Reviewer: Isso! E o submit colocar um label também.

Author: Eu acho interessante @HenriqueMorato. Você concorda @joaorsalmeida? Caso positivo, vou fazer um commit com a alteração.

Reviewer 3: Acho que quanto mais agregar na informação, melhor. 
E aí caso optemos por fazer a alteração, sugiro colocar uma orientação no manual de tradução e normalizar em todos os formulários do guia. Mas não considero um impeditivo para este PR ser mergeado. :wink:

Reviewer: @cleicar, vamos modificar então? Vou fazer a sugestão para facilitar :)

Reviewer: Ainda não pode no commentar nessas linhas tem até uma [issue](https://github.com/isaacs/github/issues/284) aberta pra isso. Você pode alterar @cleicar ?

Author: Acabei de subir a alteração! Obrigada :)

<% end %>
```

Se você atualizar a página, você vai ver o mesmo formulário do nosso exemplo acima.

> Construir formulários no Rails é tão simples assim!

Reviewer: ```suggestion
Construir formulários no Rails é simples assim!
```

Acho que tirando o `tão` essa frase parece soar melhor, mas vou pedir ajuda dos aniversários do @campuscode/guia-rails-team. O que você acha @cleicar ?

Reviewer 2: 👍 @HenriqueMorato

Quando você chama o `form_with`, você passa para ele a identificação do escopo para esse formulário.

> Neste caso, o *symbol* `:article`. Isso mostra ao `form_with` para qual escopo é o formulário em questão. Dentro do bloco deste método, o objeto `FormBuilder` - representado pelo `form` - é usado para construir dois títulos e dois campos textos, para cada título e texto de um artigo. Finalmente, uma chamada ao `submit` do objeto `form` criará um botão de submissão do formulário.

Reviewer: ```suggestion
Neste caso, o *symbol* `:article`. Isso mostra ao `form_with` para qual escopo é o formulário em questão. Dentro do bloco deste método, o objeto `FormBuilder` - representado pelo `form` - é usado para construir dois títulos e dois campos textos, um para cada título (*labels*) e texto (*text fields*) de um artigo. Finalmente, uma chamada ao `submit` do objeto `form` criará um botão de submissão do formulário.
```

Estou na dúvida mas talvez podemos deixar o termo em inglês para fazer mais alusão ao código, o que acha @cleicar ?

> Neste caso, o *symbol* `:article`. Isso mostra ao `form_with` para qual escopo é o formulário em questão. Dentro do bloco deste método, o objeto `FormBuilder` - representado pelo `form` - é usado para construir dois titulos e dois campos textos, para cada titulo e texto de um artigo. Finalmente, uma chamada ao `submit` do objeto `form` irá criar um botão de submissão do formulário.

Reviewer 3: ```suggestion
Neste caso, o *symbol* `:article`. Isso mostra ao `form_with` para qual escopo é o formulário em questão. Dentro do bloco deste método, o objeto `FormBuilder` - representado pelo `form` - é usado para construir dois títulos e dois campos textos, para cada título e texto de um artigo. Finalmente, uma chamada ao `submit` do objeto `form` criará um botão de submissão do formulário.
```

> Porém, existe um problema com este formulário. Se você inspecionar o HTML que foi gerado, visualizando o código-fonte da página, você verá que o atributo *action* do formulário está apontando para `/articles/new`. Isso é um problema porque esta rota vai exatamente para a mesma página que você está no momento, e essa rota deveria ser usada apenas para exibir o formulário para um novo artigo.

Reviewer: ```suggestion
Porém, existe um problema com este formulário. Se você inspecionar o HTML que foi gerado, visualizando o código-fonte da página, você verá que o atributo `action` do formulário está apontando para `/articles/new`. Isso é um problema porque esta rota vai exatamente para a mesma página que você está no momento, e essa rota deveria ser usada apenas para exibir o formulário para um novo artigo.
```

Aqui acho que seria legal manter como código, o que você acha @cleicar ?

> Porém, existe um problema com este formulário. Se você inspecionar o HTML que foi gerado, visualizando o código-fonte da págia, você vai ver que o atributo *action* do formulário está apontando para `/articles/new`. Isso é um problema porque esta rota vai exatamente para a mesma página que você está no momento, e essa rota deveria ser usada apenas para exibir o formulário para um novo artigo.

Reviewer 3: ```suggestion
Porém, existe um problema com este formulário. Se você inspecionar o HTML que foi gerado, visualizando o código-fonte da página, você verá que o atributo *action* do formulário está apontando para `/articles/new`. Isso é um problema porque esta rota vai exatamente para a mesma página que você está no momento, e essa rota deveria ser usada apenas para exibir o formulário para um novo artigo.
```

> O formulário precisa usar uma URL diferente para que consigar ir para outro lugar. Isso pode ser feito de forma simples com a opção `:url` do `form_with`. Normalment no Rails, a *action* que é usada para novas submissões de formários é chamada "create", então o formulário precisa apontar para esta ação.

Reviewer 3: ```suggestion
O formulário precisa usar uma URL diferente para que consigar ir para outro lugar. Isso pode ser feito de forma simples com a opção `:url` do `form_with`. Normalmente, no Rails, a *action* que é usada para novas submissões de formulários é chamada "create", então o formulário precisa apontar para esta ação.
```

Altere a linha do `form_with` dentro de `app/views/articles/new.html.erb` para ficar da seguinte forma:

```html+erb
<%= form_with scope: :article, url: articles_path, local: true do |form| %>
```

Neste exemplo, o *helper* `articles_path` é passado para a opção `:url`.

> 

Reviewer: Acho que aqui ficou uma linha a mais.

```suggestion
```

         root GET    /                            welcome#index
```

> O *helper* `articles_path` diz ao Rails para pontar o formulário para o padrão de URI associado com o prefixo `articles`; e o formulário vai (por padrão) enviar uma requisição `POST` para esta rota. Isto esta associado com a *action* `create` do `controller` atual, no caso o `ArticlesController`.

Reviewer: ```suggestion
O *helper* `articles_path` diz ao Rails para apontar o formulário para o padrão de URI associado com o prefixo `articles`; e o formulário vai (por padrão) enviar uma requisição `POST` para esta rota. Isto esta associado com a *action* `create` do `controller` atual, no caso o `ArticlesController`.
```

Com o formulário e sua rota definidos, você será capaz de preencher o formulário e clicar no botão *submit* para iniciar o processo de criação de um novo artigo, então vá em frente e faça isso. Quando você enviar, você verá um erro familiar:

![Unknown action create for ArticlesController]
(images/getting_started/unknown_action_create_for_articles.png)

> Agora você precisa adicionar a *action* create dentro do `ArticlesController` para que isso funcione.

Reviewer 3: ```suggestion
Agora você precisa adicionar a *action* `create` dentro do `ArticlesController` para que isso funcione.
```

> NOTE: Por padrão o `form_with` submete formulários usando Ajax, evitando assim o redirecionamento da página inteira. Para tornar este guia mais fácil de entender, por hora, nós desabilitamos isso ao utilizar o `local: true`.

Reviewer 3: ```suggestion
NOTE: Por padrão o `form_with` submete formulários usando Ajax, evitando assim o redirecionamento da página inteira. Para tornar este guia mais fácil de entender, por ora, nós desabilitamos isso ao utilizar o `local: true`.
```
